pkg/webhook/authorization/http.go

```diff
@@ -0,0 +1,165 @@
+	var body []byte
+	var err error
+	ctx := r.Context()
+	if wh.WithContextFunc != nil {
+		ctx = wh.WithContextFunc(ctx, r)
+	}
+
+	var reviewResponse Response
+	if r.Body == nil {
+		err = errors.New("request body is empty")
+		wh.log.Error(err, "bad request")
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+
+	defer r.Body.Close()
+	if body, err = ioutil.ReadAll(r.Body); err != nil {
+		wh.log.Error(err, "unable to read the body from the incoming request")
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		err = fmt.Errorf("contentType=%s, expected application/json", contentType)
+		wh.log.Error(err, "unable to process a request with an unknown content type", "content type", contentType)
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+
+	// Decode request body into authorizationv1.SubjectAccessReviewSpec structure
+	sar, actualTokRevGVK, err := wh.decodeRequestBody(body)
+	if err != nil {
+		wh.log.Error(err, "unable to decode the request")
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+	wh.log.V(1).Info("received request", "UID", sar.UID, "kind", sar.Kind)
+
+	reviewResponse = wh.Handle(ctx, Request{sar.SubjectAccessReview})
+	wh.writeResponseTyped(w, reviewResponse, actualTokRevGVK)
+}
+
+// writeResponse writes response to w generically, i.e. without encoding GVK information.
+func (wh *Webhook) writeResponse(w io.Writer, response Response) {
+	wh.writeSubjectAccessReviewResponse(w, response.SubjectAccessReview)
+}
+
+// writeResponseTyped writes response to w with GVK set to subjRevGVK, which is necessary
+// if multiple SubjectAccessReview versions are permitted by the webhook.
+func (wh *Webhook) writeResponseTyped(w io.Writer, response Response, subjRevGVK *schema.GroupVersionKind) {
+	ar := response.SubjectAccessReview
+
+	// Default to a v1 SubjectAccessReview, otherwise the API server may not recognize the request
+	// if multiple SubjectAccessReview versions are permitted by the webhook config.
+	if subjRevGVK == nil || *subjRevGVK == (schema.GroupVersionKind{}) {
+		ar.SetGroupVersionKind(authorizationv1.SchemeGroupVersion.WithKind("SubjectAccessReview"))
+	} else {
+		ar.SetGroupVersionKind(*subjRevGVK)
+	}
+	wh.writeSubjectAccessReviewResponse(w, ar)
+}
+
+// writeSubjectAccessReviewResponse writes ar to w.
+func (wh *Webhook) writeSubjectAccessReviewResponse(w io.Writer, ar authorizationv1.SubjectAccessReview) {
+	if err := json.NewEncoder(w).Encode(ar); err != nil {
+		wh.log.Error(err, "unable to encode the response")
+		wh.writeResponse(w, Errored(err))
+	}
+	res := ar
+	if log := wh.log; log.V(1).Enabled() {
+		log.V(1).Info("wrote response", "UID", res.UID, "authorized", res.Status.Allowed)
+	}
```

Comment:
Any reason for all this?
```suggestion
	wh.log.V(1).Info("wrote response", "UID", ar.UID, "authorized", ar.Status.Allowed)
```

Reply:
For cross-referencing is this what we do in other webhook HTTP handlers, I believe it's cause technically this should work without a logger.
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/admission/http.go#L134-L139
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/authentication/http.go#L140-L142

At least that was my thinking when I kept it in the `authentication` webhook.

Reply:
I'm pretty sure `Enabled()` exists so you can write another action to take if the logger at some level is not turned on. If you do `wh.log.V(notEnabledInt).Info(...)` the log message won't be written.

Reply:
TIL! Thanks, that's super good to know. there seems to be a couple of these places we should/can clean up throughout the entire `pkg/webhook/...` packages then. For the others should we file some issues to make sure they all get cleaned up?

Reply:
Do you want me to change this for `authorization` in this PR? Or should we proceed as is for now so that all three occasions (admission, authentication, authorization) can be cleaned up in one go?

Reply:
IMO the less changes later, the better (do it here).

Reply:
done with 2767220 (for all webhook handlers)